sdk/helper/metricregistry/metricregistry_test.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package metricregistry
+
+import (
+	"testing"
+
+	promsink "github.com/armon/go-metrics/prometheus"
+	"github.com/stretchr/testify/require"
+)
+
+var testGauges = []GaugeDefinition{
+	{
+		Name: []string{"test_gauge"},
+		Help: "A test gauge",
+	},
+	{
+		Name: []string{"test_gauge2"},
+		Help: "Another test gauge",
+	},
+}
+
+var testCounters = []CounterDefinition{
+	{
+		Name: []string{"test_counter"},
+		Help: "A test counter",
+	},
+	{
+		Name: []string{"test_counter2"},
+		Help: "Another test counter",
+	},
+}
+
+var testSummaries = []SummaryDefinition{
+	{
+		Name: []string{"test_summary"},
+		Help: "A test summary",
+	},
+	{
+		Name: []string{"test_summary2"},
+		Help: "Another test summary",
+	},
+}
+
+func TestMetricRegistry(t *testing.T) {
+	// Register some metrics
+	RegisterGauges(testGauges)
+	RegisterCounters(testCounters)
+	RegisterSummaries(testSummaries)
+
+	var opts promsink.PrometheusOpts
+
+	// Add some pre-existing metrics to ensure merge is really a merge
+	opts.GaugeDefinitions = []promsink.GaugeDefinition{
+		{
+			Name: []string{"preexisting_gauge"},
+			Help: "A pre-existing gauge",
+		},
+	}
+	opts.CounterDefinitions = []promsink.CounterDefinition{
+		{
+			Name: []string{"preexisting_counter"},
+			Help: "A pre-existing counter",
+		},
+	}
+	opts.SummaryDefinitions = []promsink.SummaryDefinition{
+		{
+			Name: []string{"preexisting_summary"},
```

Comment:
What happens if one of the preexisting names clashes with one of the incoming names? Is that worth testing and/or documenting?

Reply:
Good Q. I think the answer is last-write-wins based on:

https://github.com/hashicorp/go-metrics/blob/0ec744010d013e2ce8d0d71c69a9c43bfe523efc/prometheus/prometheus.go#L209-L221

They are inserted into a `sync.Map` keyed by the hash of the metric name and constant labels so if those collide the second registration wins.

We could de-dupe or panic or something here as it's a bug to have multiple definitions, but it also doesn't do any harm since by definition both registrations are identical!

Reply:
A more subtle question is what happens if you register a metric with the same name under different types - I think the answer is still last write wins but now it will impact the actual value of the metric. That said, even before this change it's possible to name two different metrics the same way and you'll get broken and unpredictable metrics behaviour already from doing so with no warning, so I don't think this makes anything any worse...

Reply:
@raskchanky do you think it's important to add tests for this case? We don't have tests or runtime checks that code doesn't duplicate metric names in general now which would have a similar impact so I feel like it's not going to really tell us much but open to adding something if I'm missing something?